### ISSUE TITLE
need a control that combines keypad, value display and "enter" button

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/scenery-phet#271]
NumberKeypad does not have a value display or "enter" button, and every client that has used NumberKeypad has had to add these features.

In https://github.com/phetsims/sun/issues/212#issuecomment-254903023, @jbphet said:

> In response to https://github.com/phetsims/unit-rates/issues/130#issuecomment-254684114, I think we should keep NumberKeypad 'pure' and not add a display or an entry button.  Even though we have no usages now without some sort of 'Enter' button, I could easily envision a situation in the future where we want this.  I would therefore favor the idea of a subtype with the enter button and display, but I think that is essentially what NumberEntryControl and KeypadPanelNode are (not subtypes, strictly speaking, but controls that put the keypad, display, and enter button together through composition).

> I just looked at unit-rates.KeypadPanelNode, and I think @jonathanolson may have been discouraged from combining this with the others due to some differences in the API for setting and clearing listeners.  Looking at it (but not too deeply), I still feel like these could all be consolidated.  In other words, it seems like we could have one control with a readout, keypad, and enter button that could work for area-builder, making-tens, and unit rates and possibly (as @pixelzoom noted above) arithmetic and projectile-motion too.

So we'll probably add these features by composition, possibly using NumberEntryControl as a starting point.

Assigned to @ariel-phet to prioritize and assign.

**Comment by pixelzoom** [comment]
@ariel-phet As noted in https://github.com/phetsims/unit-rates/issues/130#issuecomment-255091660, I have no plans to address this in unit-rates.  There are at least 3 other sims that pre-date unit-rates and make-a-ten, and any one of them could have tackled generalization.  Recommended to assign this issue to the author of one of those sims (area-builder, arithmetic projectile-motion), or to @jonathanolson since he make an admirable attempt at this for make-a-ten.

**Comment by ariel-phet** [comment]
Ideal to do as part of the work for publishing make-a-ten

**Comment by ariel-phet** [comment]
@jonathanolson marking for dev meeting just so we can solidify specs as a group.

**Comment by pixelzoom** [comment]
Things that the API probably needs:

Value field above keypad, options for {Color|string} background color, {Color} background stroke, {Color|string} value color, {Font} value font, {Dimension2} size of the field,...

"Enter" button below keypad, options for {Node} button label, {string} alignment (left, right, center), {number} ySpacing (from bottom of keypad), {Color|string} button color, ... (maybe all button options?)

NumberKeypad: all options

**Comment by jonathanolson** [comment]
Some concerns to address about this refactoring:

Encapsulation:

- Visibility and pickability of the button need to be handled independently of the visibility and pickability of the keypad in multiple sims (requires button and keypad/readout references)
- unit-rates dynamically changes button listeners (requires button)
- Button "enabled" also needs to be accessed.

Layout:

- arithmetic/area-builder lays out the button separately from the keypad (not as a unit), generally aligning with something else.
- area-builder lays out multiple buttons (gameControlButtons) in the same spot and toggles visibility, instead of just having one "Check" button we need to worry about. 

These concerns would have me advocate for the button and keypad/readout to be handled separately, as we can't assume we can control the layout of the button, type of the button (area-builder uses multiple buttons), there is no shared button behavior/string/font/max-width/padding. The common factor between buttons is the color (F2E916) and corner-radius (4).

I'll discuss with @pixelzoom to see if there are a subset of cases (or all cases) that we can handle and factor out nicely.

That said, I believe unit-rates' KeypadPanelNode could be refactored to use NumberEntryControl, as there are multiple benefits of shared code: sizing and positioning the readout (from maxDigits), and hooking up the text readout to the value.

**Comment by pixelzoom** [comment]
@jonathanolson and I discussed via Skype.  He'd like to omit the button.  I'm OK with that, but...  I  haven't even used NumberKeypad yet (I inherited it in unit-rates), it was written and used mostly by @jbphet . @jbphet and I were asked at last week's dev meeting to weigh in on what we'd like to see in the API.  My feedback was in https://github.com/phetsims/scenery-phet/issues/271#issuecomment-256737547; I don't see @jbphet's feedback yet.  So I'm recommended that @jonathanolson should discuss with @jbphet.

**Comment by jonathanolson** [comment]
@jbphet, let me know if you have some time to discuss if we should factor out the buttons into common code.

**Comment by jonathanolson** [comment]
I'll plan to tackle KeypadPanelNode, and if in the future we have more consistent button layout/behavior, we could consider factoring out that.

**Comment by pixelzoom** [comment]
As I look more closely at NumberKeypad (trying to use it in unit-rates), I'm pretty convinced that it's not adequately generalized.  When working on KeypadPanelNode, it would be better to fix deficiencies in NumberKeypad (e.g. #272, #273, #274,...), rather than work around them in  KeypadPanelNode.

**Comment by jonathanolson** [comment]
Unassigning until NumberKeypad work is resolved, to see if there's still shared code.

**Comment by pixelzoom** [comment]
@ariel-phet FYI, the lack of closure on this issue is resulting in work-around solutions that are being copy-pasted between sims.  See https://github.com/phetsims/projectile-motion/issues/118

**Comment by jonathanolson** [comment]
Presumably implementations that are similar should be factored out (I don't copy-paste things like that).

Area Model is not a candidate for refactoring (RichText, custom readout size that can change, etc.), but Make a Ten's keypad definitely could be used (as it uses NumberEntryControl).

**Comment by pixelzoom** [comment]
Before doing this, we should discuss layout & option requirements with designers.  In the handful of sims where I could have used this component, the requirements have been different.  Can they be standardized?

**Comment by zepumph** [comment]
Potential (horrible) names for this thing:
* KeypadDisplay
* DisplayEnterKeypad
* KeypadEnterNode
* KeypadComposite

**Comment by zepumph** [comment]
I created https://github.com/phetsims/scenery-phet/issues/663 to deal with the modal dialog type solution needed for make a ten and Unit rates and projectile motion, please see https://github.com/phetsims/scenery-phet/issues/663. 

Thus it seems like this issue is really more about adding the composition of numbers display bar and enter button. If I end up being wrong, and really this is all one type, then we can consolidate https://github.com/phetsims/scenery-phet/issues/663 back into this issue.

From https://github.com/phetsims/scenery-phet/issues/271#issuecomment-257743382, @pixelzoom said:
>. . .  it would be better to fix deficiencies in NumberKeypad (e.g. #272, #273, #274,...), rather than work around them in KeypadPanelNode.

I see that only #272 has yet to be addressed, and I will take a look over there. In addition, I will first go through any loose ends in https://github.com/phetsims/scenery-phet/issues/283 to make sure Keypad is in tip-top shape before trying to exercise it in new common code types.

**Comment by jbphet** [comment]
@zepumph is going to take this on (thanks @zepumph!), I'm unassigning myself for now.

**Comment by zepumph** [comment]
Some ideas to progress this issue.

1. We should probably get some design time for consultation.

Here are some keypads around the project:

**Modal, in Projectile Motion:**
![image](https://user-images.githubusercontent.com/6856943/127057687-ef62954b-bd08-46dc-9573-20560fadff6c.png)

**Modal, in Unit Rates**
![image](https://user-images.githubusercontent.com/6856943/127057762-cd76781f-80c0-471a-adc4-0068827a0094.png)


**Non modal, in area-model-algebra**
![image](https://user-images.githubusercontent.com/6856943/127058018-8368197d-495f-4e3f-b15a-c20ff4073a3b.png)

**Modal, in Fourier**
![image](https://user-images.githubusercontent.com/6856943/127058130-5c600a68-cdf7-4939-a87e-3f123c6e2771.png)


Is this enough of a consensus to say that we can take a Node and stack from top to bottom a display, a keypad, and an enter button? If that is too specific, we could implement a "layout function" like in NumberControl, with the default of the aforementioned. I think that this could get us pretty far.

I feel a bit like I need to speak with a designer before continuing, which means that in a sense, further progress in https://github.com/phetsims/projectile-motion/issues/127 is blocked by this issue.

**Comment by zepumph** [comment]
@amanda-phet. Does it seem reasonable that our composite Keypad type generally have these three features. I think that for the most part. What other constraints might be needed by other sims? For example, the top label in a few of these sims. Presumably this would be handled by a subtype, but maybe not?

**Comment by amanda-phet** [comment]
>@amanda-phet. Does it seem reasonable that our composite Keypad type generally have these three features. I think that for the most part.

yes, those three components (display, keypad, and enter button) seem to be consistent between the keypads. It seems like we need to support custom buttons in addition to 0-9 and enter, and perhaps a label. 

Other features to consider are input types (if that's what to call it). In Area Model, we need to be able to enter a variable in addition to a number. Otherwise I can't think of any other potential features. Might be good for @arouinfar to also chime in. 

Arithmetic has a keypad-is component with a button that only becomes available once a number is entered:
![image](https://user-images.githubusercontent.com/7852916/127919163-e5c29fcb-a69e-4605-8587-3773564023c9.png)

**Comment by zepumph** [comment]
Maybe it also seems important that you can specify and adjust an optional "range text." That seems to be a need that will continue to be important to a keypad. The two usages of that above place it at the top, would it ever not be at the top?

**Comment by zepumph** [comment]
We have `SCENERY_PHET/KeypadDialog`. In my opinion. That class should compose the type that we are talking about here. It should not depend on a modal dialog wrapping it.
